Add architecture and design principles doc (#715)
nginxinc/nginx-kubernetes-gateway: docs/developer/design-principles.md

# Design Principles

> The aim of NGINX Kubernetes Gateway is to become a fundamental infrastructure component within a Kubernetes cluster,

Reviewer: ```suggestion
The aim of the NGINX Kubernetes Gateway is to become a fundamental infrastructure component within a Kubernetes cluster,
```

> serving as both an ingress and egress point for traffic directed towards the services (applications) running either

Reviewer: ```suggestion
serving as both an ingress and egress point for traffic directed towards the services (applications) running
```
Slightly redundant, I'd cut this for conciseness.

within or outside the cluster. Leveraging NGINX as a data plane technology, it harnesses the well-established reputation
of NGINX as an open-source project widely recognized for its role as a web server, proxy, load balancer, and content
cache. NGINX is renowned for its stability, high performance, security, and rich feature set, positioning it as a
critical infrastructure tool. Notably, once properly configured and operational, NGINX requires minimal attention,

> making it reliable and "boring" software.

Reviewer: I would consider removing boring and perhaps replacing it with another adjective, such as "flexible" - to emphasise the paragraph's example of uses cases.

Boring is generally a negative adjective and although many experienced engineers enjoy boring (and might refer to the "Choose Boring Technology" talk, a novice user or one less versed popular ideas & talks might be put off.

> Likewise, the goal for the NGINX Kubernetes Gateway is to embody the same qualities as NGINX and be regarded as "boring"

Reviewer: ```suggestion
The NGINX Kubernetes Gateway aims to embody the same qualities as NGINX and become familiar, trustworthy and reliable
```

Edit suggestion following my earlier suggestion regarding the use of "boring".
